runtime/registry.go

```diff
@@ -79,11 +71,139 @@ func (r *Runtime) CreateInstance(ctx context.Context, inst *drivers.Instance) er
 	return nil
 }
 
-func (r *Runtime) DeleteInstance(ctx context.Context, instanceID string) error {
-	err := r.Registry().DeleteInstance(ctx, instanceID)
+func (r *Runtime) DeleteInstance(ctx context.Context, instanceID string, dropDB bool) error {
+	inst, err := r.Registry().FindInstance(ctx, instanceID)
 	if err != nil {
+		if errors.Is(err, drivers.ErrNotFound) {
+			return nil
+		}
 		return err
 	}
 
-	return nil
+	svc, err := r.catalogCache.get(ctx, r, instanceID)
+	if err != nil { // return error if db handlers can't be opened
+		return err
+	}
+
+	// delete instance related data if catalog is not embedded
+	if !inst.EmbedCatalog {
+		err := svc.Catalog.DeleteEntries(ctx, instanceID)
+		if err != nil {
+			return err
+		}
+	}
+
+	if dropDB {
+		// ignoring the dropDB error since if db is already dropped it may not be possible to retry
+		// also the only error that os.remove returns is patherror which should be ignored as well
+		_ = svc.Olap.DropDB()
+	}
+
+	r.evictCaches(ctx, inst)
+	// delete instance
+	return r.Registry().DeleteInstance(ctx, instanceID)
+}
+
+// EditInstance edits exisiting instance.
+// Confirming to put api specs, it is expected to send entire existing instance data.
+// The API compares and only evicts caches if drivers or dsn is changed.
+// This is done to ensure that db handlers are not unnecessarily closed
+func (r *Runtime) EditInstance(ctx context.Context, inst *drivers.Instance) error {
+	olderInstance, err := r.Registry().FindInstance(ctx, inst.ID)
+	if err != nil {
+		return err
+	}
+
+	// 1. changes in olap driver or olap dsn
+	olapChanged := olderInstance.OLAPDriver != inst.OLAPDriver || olderInstance.OLAPDSN != inst.OLAPDSN
+	if olapChanged {
+		// Check OLAP connection
+		olap, _, err := r.checkOlapConnection(inst)
```

Comment:
`olap` not closed

Reply:
Not sure about a good design here, but it should either be closed or added to the conn cache

Reply:
(Same problem applies in `CreateInstance`)